server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java

```diff
@@ -0,0 +1,300 @@
+
+            return sourceShards[shardNum];
+        }
+
+        /**
+         * Set the current shard state of a source shard
+         * Currently the only legal transition is from SOURCE to DONE and any other transition will assert.
+         * This could be expressed through a markSourceDone API but this form is the same shape as {@link #setTargetShardState}
+         * and leaves the door open for additional source states.
+         * @param shardNum an index into the shards which must be no greater than the number of shards before split
+         * @param sourceShardState the state to which the shard should be set
+         */
+        public void setSourceShardState(int shardNum, SourceShardState sourceShardState) {
+            assert shardNum >= 0 && shardNum < sourceShards.length : "source shardNum is out of bounds";
+            assert sourceShards[shardNum].ordinal() + 1 == sourceShardState.ordinal() : "invalid source shard state transition";
+            assert sourceShardState == SourceShardState.DONE : "can only move source shard state to DONE";
+            for (var target : getTargetStatesFor(shardNum)) {
+                assert target == TargetShardState.DONE : "can only move source shard to DONE when all targets are DONE";
+            }
+
+            sourceShards[shardNum] = sourceShardState;
+        }
+
+        /**
+         * Get the current target state of a shard
+         * @param shardNum an index into shards greater than or equal to the old shard count and less than the new shard count
+         * @return the target shard state for the shard identified by shardNum
+         */
+        public TargetShardState getTargetShardState(int shardNum) {
+            var targetShardNum = shardNum - oldShardCount;
+
+            assert targetShardNum >= 0 && targetShardNum < targetShards.length : "target shardNum is out of bounds";
+
+            return targetShards[targetShardNum];
+        }
+
+        /**
+         * Set the target state of a shard
+         * The only legal state in the split state machine is the one following the shard's current state.
+         * The reason for this API rather than an advanceState API is to confirm that the caller knows
+         * what the current state is when setting it.
+         * @param shardNum an index into shards greater than or equal to the old shard count and less than the new shard count
+         * @param targetShardState the state to which the shard should be set
+         */
+        public void setTargetShardState(int shardNum, TargetShardState targetShardState) {
+            var targetShardNum = shardNum - oldShardCount;
+
+            assert targetShardNum >= 0 && targetShardNum < targetShards.length : "target shardNum is out of bounds";
+            assert targetShards[targetShardNum].ordinal() + 1 == targetShardState.ordinal() : "invalid target shard state transition";
+
+            targetShards[targetShardNum] = targetShardState;
```

Comment:
Similar to above, we should make this object immutable.

Reply:
will do.

Reply:
Done in 0e69a76